src/metabase/query_processor/middleware/resolve_joined_fields.clj

```diff
@@ -50,21 +50,30 @@
                              :joins      joins
                              :candidates candidate-tables}))))))))
 
+(defn- primary-source-table-id
+  "Get the ID of the 'primary' table towards which this query is pointing at: either the `:source-table` or indirectly
+  thru some number of `:source-query`s."
+  [{:keys [source-table source-query]}]
+  (or source-table
+      (when source-query
+        (recur source-query))))
```

Comment:
No possibility of infinite loop here?

Reply:
It destructures `source-query` when it recurs so the only way it could have an infinite loop would be if we had infinite source queries